### PR TITLE
Focus meta provider

### LINKED ISSUE #729: Create meta for containing focus
**Enhancement**

From the widgets focus management issue (https://github.com/dojo/widgets/issues/309#issuecomment-339100993):

> **Provide meta information about whether focus is within a given node**
>
>This can be handled like a combination of Drag and Intersection meta. Similar to Drag, it will need a document-level event listener for the focus event, and like Intersection it can invalidate when focus moves into or out of a provided node.

I imagine the API would be something like:
`public get(key: string): boolean`

### PR DESCRIPTION
**Pull request by smhigley**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Addresses #729. Creates Focus meta that provides information about whether a node is currently focused or contains focus.

Updates sinon to ^4.1.3, along with syntax updates in any tests that required it.

### CODE REVIEW

**Comment by agubler** [comment]
Can you add back the updated `package-lock.json` please?

File: package.json [hunk 1]
```diff
@@ -40,7 +40,7 @@
     "@types/jsdom": "2.0.*",
     "@types/ramda": "^0.25.5",
     "@types/selenium-webdriver": "^3.0.8",
-    "@types/sinon": "^1.16.31",
+    "@types/sinon": "^4.1.2",
```
Comment (agubler) [inline package.json]: Probably wise to pin this dependency :)

File: package.json [hunk 2]
```diff
@@ -57,7 +57,7 @@
     "ramda": "^0.25.0",
     "selenium-webdriver": "3.6.0",
     "serve": "^6.4.1",
-    "sinon": "^1.17.6",
+    "sinon": "^4.1.3",
```
Comment (agubler) [inline package.json]: And this dependency!

File: src/meta/Focus.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+		return {
+			active: node === this._activeElement,
+			containsFocus: node.contains(this._activeElement)
+		};
+	}
+
+	public set(key: string | number, selector?: string) {
+		const node = this.getNode(key);
+		const target = node && selector ? node.querySelector(selector) : node;
```
Comment (agubler) [inline src/meta/Focus.ts]: Does this break encapsulation? Allowing a parent to focus on a node from a child widget?

Comment (agubler) [reply src/meta/Focus.ts]: Also you should be able to do `as HTMLElement` here rather than on usage.

File: src/meta/Focus.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		target && (target as HTMLElement).focus();
+	}
+
+	private _onFocus() {
```
Comment (agubler) [inline src/meta/Focus.ts]: if you do 
```ts
private _onFocus = () => {
    // on focus stuff here
}
```

You shouldn't need to create the bound function.

File: src/meta/Focus.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+export class Focus extends Base {
+	private _activeElement: Element;
+	private _boundFocusHandler = this._onFocus.bind(this);
+
+	public get(key: string | number): FocusResults {
+		const node = this.getNode(key);
+
+		if (!node) {
+			return assign({}, defaultResults);
```
Comment (agubler) [inline src/meta/Focus.ts]: Do we need to use `assign` from dojo/core here? Can we just do:
```ts
return { ...defaultResults };
```

File: src/meta/Focus.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	containsFocus: boolean;
+}
+
+const defaultResults = {
+	active: false,
+	containsFocus: false
+};
+
+export class Focus extends Base {
+	private _activeElement: Element;
```
Comment (agubler) [inline src/meta/Focus.ts]: If another component uses `Focus` meta and sets focus, is the element set as `_activeElement` then incorrect? It looks like we set it once when it is not set, but if another meta sets focus, it doesn't get reset so the return from the `get` would be still return `true` for `node === this._activeElement`?

File: tests/unit/meta/Focus.ts [hunk 7]
```diff
@@ -0,0 +1,163 @@
+const { registerSuite } = intern.getInterface('object');
```
Comment (agubler) [inline tests/unit/meta/Focus.ts]: For new tests can we use the `bdd` interfaces?

**Comment by agubler** [comment]
Could you please add a section to the README for the usage?

File: src/meta/Focus.ts (hunk 5 above)
Comment (smhigley) [reply src/meta/Focus.ts]: I think I just copied this pattern from other meta providers, but `{ ...defaultResults }` looks better

File: src/meta/Focus.ts (hunk 6 above)
Comment (agubler) [reply src/meta/Focus.ts]: Right, gotcha.

File: src/meta/Focus.ts (hunk 3 above)
Comment (smhigley) [reply src/meta/Focus.ts]: Having a child selector seemed like a good way to work around `DomWrapper` or wanting to call focus on a child widget, but it does break encapsulation. One possible solution might be to check focusability and send focus to the first focusable child found, which would allow a parent to send focus inside a child widget without knowing much about its DOM.

It’s not perfect, since the parent would still be assuming that something is focusable inside the child, and checking focusability is a bit fragile. I'm not sure what a better solution would look like, however.

Comment (agubler) [reply src/meta/Focus.ts]: Why does a widget need be able to focus something that it doesn't declare? I was kinda expecting the focus meta to simply return whether a node was currently focussed (or as you have done also indicate a node currently contains the focussed node) or be able to set focus on a node that it explicitly declares using `v()`.

Comment (smhigley) [reply src/meta/Focus.ts]: Some examples: The ComboBox needs to be able to focus on its TextInput widget, Select needs to focus on the Listbox root, and Calendar needs to focus on internal CalendarCell widgets.

Outside those examples, I imagine many widgets that control focus are occasionally going to want to send focus to a form field widget, or a Button widget.

Comment (agubler) [reply src/meta/Focus.ts]: All these things are widgets that should provide an API to focus and use the focus meta internally?

Comment (smhigley) [reply src/meta/Focus.ts]: I'm not sure what that API should look like. Right now we use a rather fragile pattern where the focusable widget has `callFocus` and `onFocusCalled` properties, and it's up to the parent to set `callFocus` back to false in the `onFocusCalled` callback. If the parent doesn't handle it correctly, you risk repeatedly stealing focus and completely breaking keyboard navigation. I hope we can provide a tool that allows widget authors to move focus more easily, and with fewer pitfalls.

The basic problem is calling focus is a strictly one-time thing, so it can't be controlled easily through properties. There's also the issue that controlling focus is a parent-level concern, but needs to be done on child DOM.

Comment (matt-gadd) [reply src/meta/Focus.ts]: it does seem like we need some concept of whether something is focusable at a widget level. that way we could at least signal the intent via the parent but it would still be the child widget that sets it's own focus.

File: src/meta/Base.ts [hunk 8]
```diff
@@ -24,7 +24,7 @@ export class Base extends Destroyable implements WidgetMetaBase {
 		return this.nodeHandler.has(key);
 	}
 
-	protected getNode(key: string | number): Element | undefined {
+	public getNode(key: string | number): Element | undefined {
```
Comment (agubler) [inline src/meta/Base.ts]: Why did you need to make this `public`?

Comment (smhigley) [reply src/meta/Base.ts]: The `WebAnimation` meta stubs it in unit tests, and the updated version of sinon complained if it wasn't `public`

Comment (agubler) [reply src/meta/Base.ts]: Ah that's a little annoying :(

**Comment by smhigley** [comment]
@tomdye Focus stealing is actually more of an issue when using `focus` as a property in `render`. The way Meta currently handles it means that it should be easy to only call it once (e.g. in an event handler). If it were to be in `render`, it would be much more likely to steal if it were set to `true` on multiple renders.

The problem with having `focus` be a property is that moving focus anywhere is a single event, not an ongoing state. As soon as you try to control `focus` with properties you need to be really careful to not allow `focus` to be `true` for more than a single render. We do that now with `focus`/`onFocusCalled`, but it's fragile and the consequences for a widget user messing it up are pretty high.

File: src/meta/Base.ts (hunk 8 above)
Comment (smhigley) [reply src/meta/Base.ts]: yeah :/

**Comment by tomdye** [comment]
@smhigley from my understanding, if using a property to control focus within the `render` function, in order to call `focus()` twice, you'd need to set the property from `true -> false -> true` due to our property diff approach. So I don't think that it is such a cause for concern.

**Comment by agubler** [comment]
For now can you add the `test.skip` in each of the tests that may need to be skipped?

**Comment by agubler** [comment]
@smhigley intern has been patched to support `skip` in the `beforeEach`, but there seems to be real CI failures now.

File: src/meta/Base.ts (hunk 8 above)
Comment (tomdye) [reply src/meta/Base.ts]: I think we need to raise an issue to address the `WebAnimation` tests rather than doing this here. Should probably have bumped `sinon` in it's own PR.
